addon/components/button-spinner.ts

```diff
@@ -0,0 +1,158 @@
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+import { isNone, typeOf } from '@ember/utils';
+import { cancel, later } from '@ember/runloop';
+import { EmberRunTimer } from '@ember/runloop/types';
+import { resolve } from 'rsvp';
+import { ButtonArgs } from '@gavant/ember-button-basic/components/button';
+
+export interface ButtonSpinnerArgs extends ButtonArgs {
+    isSpinning?: boolean;
+    showSuccess?: boolean;
+    showError?: boolean;
+    loadingClass?: string;
+    successClass?: string;
+    successAnimationClass?: string;
+    successIcon?: string;
+    successIconClass?: string;
+    action?: (event: Event) => Promise<any>;
+}
+
+export default class ButtonSpinner extends Component<ButtonSpinnerArgs> {
+    @tracked isSpinning: boolean = false;
+    @tracked successShown: boolean = false;
+    @tracked errorShown: boolean = false;
+
+    successStateDuration: number = 3000;
+    successAnimateInDuration: number = 1000;
+    errorStateDuration: number = 1000;
+    errorAnimateInDuration: number = 1000;
+    showResultTimer: EmberRunTimer | null = null;
+
+    get buttonType() {
+        //default the button `type` attribute to 'submit', as spinner buttons
+        //are usually used for form submission
+        return this.args.buttonType || 'submit';
+    }
+
+    get preventDefault() {
+        //default the button action to prevent the default behavior, as spinner buttons
+        //are usually used for form submission
+        return !isNone(this.args.preventDefault) ? this.args.preventDefault : true;
```

Comment:
could you use this.args?.preventDefault? instead of isNone?

Reply:
no the `?` for optional chaining is just for conditionally accessing deeper properties, it wont actually return true/false here like `isNone()` does.

I could technically write this as:
```
return this.args.preventDefault ?? true;
```
But I believe thats a newer typescript syntax and VSCode currently has issues applying syntax highlighting it, leading to mangled highlighting and grossness for the rest of the file, so I opted for the more verbose code for now. ^_^